{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "summary = []\n",
    "summary.append({\n",
    "    \"GWAS\": \"GCST90450224\", \n",
    "    \"leads\": \"example_data/example_leads.tsv\", \n",
    "    \"GWAS_path\": \"example_data/GCST90450224.tsv\",\n",
    "    \"chr_col\": \"chromosome\",\n",
    "    \"pos_col\": \"base_pair_location\",\n",
    "    \"maf_col\": \"effect_allele_frequency\",\n",
    "    \"beta_col\": \"beta\",\n",
    "    \"se_col\": \"standard_error\",\n",
    "    \"file_type\": \"tsv\",\n",
    "    \"ref_col\": \"other_allele\",\n",
    "    \"alt_col\": \"effect_allele\"\n",
    "    })\n",
    "\n",
    "pd.DataFrame(summary).to_csv(\"example_gwas_table.tsv\",sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "summary = []\n",
    "summary.append({\n",
    "    \"id\": \"QTD000186\", \n",
    "    \"path\": \"example_data/QTD000186.lbf_variable.txt.gz.tsv\",\n",
    "    })\n",
    "\n",
    "pd.DataFrame(summary).to_csv(\"example_eQTL_table.tsv\",sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>neg_log_10_p_value</th>\n",
       "      <th>rsid</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>info</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>285339</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.020926</td>\n",
       "      <td>0.199508</td>\n",
       "      <td>0.014613</td>\n",
       "      <td>0.037884</td>\n",
       "      <td>rs879732627</td>\n",
       "      <td>1_285339_C_A</td>\n",
       "      <td>0.662782</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>631712</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.161614</td>\n",
       "      <td>0.154556</td>\n",
       "      <td>0.017865</td>\n",
       "      <td>0.529126</td>\n",
       "      <td>rs9326622</td>\n",
       "      <td>1_631712_T_C</td>\n",
       "      <td>0.851058</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>631739</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.161388</td>\n",
       "      <td>0.154579</td>\n",
       "      <td>0.017789</td>\n",
       "      <td>0.528031</td>\n",
       "      <td>rs9283153</td>\n",
       "      <td>1_631739_A_C</td>\n",
       "      <td>0.854172</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>656627</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.147961</td>\n",
       "      <td>0.125803</td>\n",
       "      <td>0.039830</td>\n",
       "      <td>0.620621</td>\n",
       "      <td>rs771664146</td>\n",
       "      <td>1_656627_G_A</td>\n",
       "      <td>0.661197</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>656671</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.022934</td>\n",
       "      <td>0.175279</td>\n",
       "      <td>0.021045</td>\n",
       "      <td>0.047740</td>\n",
       "      <td>rs10218594</td>\n",
       "      <td>1_656671_C_T</td>\n",
       "      <td>0.609434</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787942</th>\n",
       "      <td>1</td>\n",
       "      <td>248934255</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.131715</td>\n",
       "      <td>0.175541</td>\n",
       "      <td>0.013527</td>\n",
       "      <td>0.343852</td>\n",
       "      <td>rs138689967</td>\n",
       "      <td>1_248934255_C_A</td>\n",
       "      <td>0.953348</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787943</th>\n",
       "      <td>1</td>\n",
       "      <td>248934257</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.114116</td>\n",
       "      <td>0.099312</td>\n",
       "      <td>0.046349</td>\n",
       "      <td>0.601138</td>\n",
       "      <td>rs199655021</td>\n",
       "      <td>1_248934257_A_C</td>\n",
       "      <td>0.903370</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787944</th>\n",
       "      <td>1</td>\n",
       "      <td>248936569</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.047267</td>\n",
       "      <td>0.104786</td>\n",
       "      <td>0.038687</td>\n",
       "      <td>0.185798</td>\n",
       "      <td>rs532053890</td>\n",
       "      <td>1_248936569_C_T</td>\n",
       "      <td>0.877370</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787945</th>\n",
       "      <td>1</td>\n",
       "      <td>248936926</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.088907</td>\n",
       "      <td>0.144758</td>\n",
       "      <td>0.019251</td>\n",
       "      <td>0.268333</td>\n",
       "      <td>rs199758136</td>\n",
       "      <td>1_248936926_C_G</td>\n",
       "      <td>0.948789</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>787946</th>\n",
       "      <td>1</td>\n",
       "      <td>248945104</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.068690</td>\n",
       "      <td>0.157727</td>\n",
       "      <td>0.018739</td>\n",
       "      <td>0.178355</td>\n",
       "      <td>rs573070055</td>\n",
       "      <td>1_248945104_C_T</td>\n",
       "      <td>0.789124</td>\n",
       "      <td>OA</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>787947 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chromosome  base_pair_location effect_allele other_allele      beta  \\\n",
       "0                1              285339             A            C -0.020926   \n",
       "1                1              631712             C            T  0.161614   \n",
       "2                1              631739             C            A  0.161388   \n",
       "3                1              656627             A            G -0.147961   \n",
       "4                1              656671             T            C  0.022934   \n",
       "...            ...                 ...           ...          ...       ...   \n",
       "787942           1           248934255             A            C  0.131715   \n",
       "787943           1           248934257             C            A -0.114116   \n",
       "787944           1           248936569             T            C -0.047267   \n",
       "787945           1           248936926             G            C -0.088907   \n",
       "787946           1           248945104             T            C -0.068690   \n",
       "\n",
       "        standard_error  effect_allele_frequency  neg_log_10_p_value  \\\n",
       "0             0.199508                 0.014613            0.037884   \n",
       "1             0.154556                 0.017865            0.529126   \n",
       "2             0.154579                 0.017789            0.528031   \n",
       "3             0.125803                 0.039830            0.620621   \n",
       "4             0.175279                 0.021045            0.047740   \n",
       "...                ...                      ...                 ...   \n",
       "787942        0.175541                 0.013527            0.343852   \n",
       "787943        0.099312                 0.046349            0.601138   \n",
       "787944        0.104786                 0.038687            0.185798   \n",
       "787945        0.144758                 0.019251            0.268333   \n",
       "787946        0.157727                 0.018739            0.178355   \n",
       "\n",
       "               rsid       variant_id      info ref_allele    n  \n",
       "0       rs879732627     1_285339_C_A  0.662782         OA  928  \n",
       "1         rs9326622     1_631712_T_C  0.851058         OA  928  \n",
       "2         rs9283153     1_631739_A_C  0.854172         OA  928  \n",
       "3       rs771664146     1_656627_G_A  0.661197         OA  928  \n",
       "4        rs10218594     1_656671_C_T  0.609434         OA  928  \n",
       "...             ...              ...       ...        ...  ...  \n",
       "787942  rs138689967  1_248934255_C_A  0.953348         OA  928  \n",
       "787943  rs199655021  1_248934257_A_C  0.903370         OA  928  \n",
       "787944  rs532053890  1_248936569_C_T  0.877370         OA  928  \n",
       "787945  rs199758136  1_248936926_C_G  0.948789         OA  928  \n",
       "787946  rs573070055  1_248945104_C_T  0.789124         OA  928  \n",
       "\n",
       "[787947 rows x 13 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"GCST90450224.tsv\", sep=\"\\t\")\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = df[df[\"effect_allele_frequency\"]>=0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>lbf</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_14826873_G_A</td>\n",
       "      <td>6.792095</td>\n",
       "      <td>14826873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_76433270_G_A</td>\n",
       "      <td>6.541357</td>\n",
       "      <td>76433270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_85516585_T_C</td>\n",
       "      <td>6.163146</td>\n",
       "      <td>85516585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_85517720_A_G</td>\n",
       "      <td>6.057619</td>\n",
       "      <td>85517720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_85543257_T_C</td>\n",
       "      <td>6.284835</td>\n",
       "      <td>85543257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1_218585370_G_A</td>\n",
       "      <td>6.246734</td>\n",
       "      <td>218585370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1_218588339_C_T</td>\n",
       "      <td>7.502817</td>\n",
       "      <td>218588339</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              variant       lbf   position\n",
       "0   chr1_14826873_G_A  6.792095   14826873\n",
       "1   chr1_76433270_G_A  6.541357   76433270\n",
       "2   chr1_85516585_T_C  6.163146   85516585\n",
       "3   chr1_85517720_A_G  6.057619   85517720\n",
       "4   chr1_85543257_T_C  6.284835   85543257\n",
       "5  chr1_218585370_G_A  6.246734  218585370\n",
       "6  chr1_218588339_C_T  7.502817  218588339"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "def approx_bf_estimates(variant, z, V, type, suffix=None, sdY=1, effect_priors={'quant': 0.15, 'cc': 0.2}):\n",
    "    sd_prior = effect_priors['quant'] * sdY if type == \"quant\" else effect_priors['cc']\n",
    "    \n",
    "    r = sd_prior**2 / (sd_prior**2 + V)\n",
    "    lABF = 0.5 * (np.log(1 - r) + (r * z**2))\n",
    "    \n",
    "    ret = pd.DataFrame({\"variant\": variant, 'lbf': lABF})\n",
    "    \n",
    "    if suffix is not None:\n",
    "        ret.columns = [f\"{col}.{suffix}\" for col in ret.columns]\n",
    "\n",
    "    ret.loc[:, 'lbf'] = pd.to_numeric(ret['lbf'], errors='coerce')\n",
    "\n",
    "    ret_mat = pd.DataFrame(ret.set_index('variant')['lbf']).T\n",
    "\n",
    "    return ret_mat\n",
    "\n",
    "df[\"variant\"] = (\n",
    "    \"chr\" + \"1\" + \"_\" +\n",
    "    df[\"base_pair_location\"].astype(int).astype(str) + \"_\" +\n",
    "    df[\"other_allele\"].astype(str) + \"_\" +\n",
    "    df[\"effect_allele\"].astype(str)\n",
    ")\n",
    "\n",
    "df[\"z\"] = df[\"beta\"] / df[\"standard_error\"]\n",
    "df[\"V\"] = df[\"standard_error\"] ** 2\n",
    "\n",
    "mat = approx_bf_estimates(df[\"variant\"],df[\"z\"], df[\"V\"], \"quant\") \n",
    "mat1 = mat.T\n",
    "leads = mat1[mat1[\"lbf\"]>=6].reset_index()\n",
    "leads[\"position\"] = leads[\"variant\"].astype(str).str.split(\"_\").str[1].astype(int)\n",
    "leads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 14826873,  76433270,  85516585,  85517720,  85543257, 218585370,\n",
       "       218588339])"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leads[\"position\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "leads_df= df[df[\"base_pair_location\"].isin(leads[\"position\"].unique())]\n",
    "leads_df\n",
    "\n",
    "leads_df.to_csv(\"example_leads.tsv\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
